### PR TITLE
[profiler] Add execution_trace_observer as an optional argument to profiler

### PR DESCRIPTION
**Pull request by briancoutinho**
# Update Profiler API to collect Execution Traces

## TLDR
We would like to simplify collecting Execution Trace and Kineto together. Execution Trace and Kineto both provide meaningful information that can be combined to enable benchmarking, performance analysis and simulating new hardware.
```
import torch

def main():
    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ],
        …
        excution_trace_observer=ExecutionTraceObserver() # <<<<<<< NEW
    ) as prof:
        ...
        prof.step()
```

See test/profiler/test_profiler.py 'test_execution_trace_with_kineto' for an example of using this API.

## What are Execution Traces?
[Chakra Execution Traces](https://github.com/mlcommons/chakra/wiki) offer a graph based representation of AI/ML workloads.  It stands apart from conventional AI/ML frameworks by focusing on replay benchmarks, simulators, and emulators, prioritizing agile performance modeling and adaptable methodologies.
- Chakra is part of ML Commons industry standard and is being adopted by other companies besides NVIDIA too.
- At Meta we have instrumented PyPer framework to collect Execution Traces. More details on our [PyTorch implementation of Chakra can be found here](https://github.com/mlcommons/chakra/wiki)

Chakra essentially enables benchmarking and co-design for ML Models without having to reproduce entier software stacks and helps companies collaborate together [[chakra paper](https://arxiv.org/pdf/2305.14516.pdf)]

## Why correlate Execution Trace with PyTorch/Kineto Trace

Both Execution Traces and Kineto/ provide different types of information and combining. While PyTorch ETs focus on CPU operators with explicit dependencies between them, Kineto traces encode GPU operators with their start and end times. In addition, collecting them at different timestamps will be inaccurate as several operations (NCCL, Embedding lookup) are data dependent and may not match correctly.
Thus, it makes sense to collect both ET and Kineto together. The problem is that there are two code paths. 

## Proposal
The proposal is to modify the PyTorch profiler (Kineto) API to enable execution trace to be collected simultaneously, see TLDR section

# Testing
Updated the unit test for collecting kineto and Execution Trace together. 
- Check the collected ET has right range of events.
- Compare two sets of IDs - record func Ids in ET and external IDs in Kineto. We check if these have a constant difference.

```
pytest test/profiler/test_profiler.py  -k test_execution_trace_with_kineto -rP

Running 1 items in this shard

test/profiler/test_profiler.py [W execution_trace_observer.cpp:682] Enabling Execution Trace Observer
STAGE:2024-03-05 09:05:05 1119546:1119546 ActivityProfilerController.cpp:314] Completed Stage: Warm Up
[W execution_trace_observer.cpp:694] Disabling Execution Trace Observer
STAGE:2024-03-05 09:05:05 1119546:1119546 ActivityProfilerController.cpp:320] Completed Stage: Collection
STAGE:2024-03-05 09:05:05 1119546:1119546 ActivityProfilerController.cpp:324] Completed Stage: Post Processing
```

### CODE REVIEW

File: torch/profiler/profiler.py [hunk 1]
```diff
@@ -46,6 +46,95 @@ def supported_activities():
...
+        self._registered = False
+        self._execution_trace_running = False
+
+    def __del__(self):
+        """
+        Calls unregister_callback() to make sure to finalize outputs.
+        """
+        self.unregister_callback()
+
+    def register_callback(self, output_file_path: str) -> Self:
```
Comment (briancoutinho) [inline torch/profiler/profiler.py:77]: Note one minor change is a returned reference to Self!

File: torch/profiler/profiler.py [hunk 2]
```diff
@@ -46,6 +46,95 @@ def supported_activities():
     return torch.autograd._supported_activities()
 
 
+class ExecutionTraceObserver:
```
Comment (briancoutinho) [inline torch/profiler/profiler.py:49]: This class was moved up so we can refer to it in other classes

**Comment by srinivas212** [comment]
Thanks for this PR, @briancoutinho. It is much appreciated. This will be extremely useful for numerous simulation and co-design for future systems. Like you clearly articulated - we need both ET and kineto for these use cases. We are currently jumping through hoops to collect kineto and ET together and running into numerous reliability issues.

**Review by shengfukevin — COMMENTED** [review]
Why rf_ids in execution trace and external_ids in Kineto trace have the constant delta? Can we make them the same since they start and stop simultaneously?

File: test/profiler/test_profiler.py [hunk 3]
```diff
@@ -354,34 +377,41 @@ def trace_handler(p):
...
+
+        # Uncomment for debugging
+        # print("Output kineto = ", kt.name)
+        # print("Output ET = ", et.get_output_file_path())
+
+        self.assertEqual(trace_called_num, 1)
+        self.assertEqual(fp.name, et.get_output_file_path())
 
         # cleanup
         et.unregister_callback()
```
Comment (sanrise) [inline test/profiler/test_profiler.py:413]: We can unregister inside the context manager's exit/stop method (https://github.com/pytorch/pytorch/blob/main/torch/profiler/profiler.py#L635) with something like this in there:

```
if self.et: 
  et.unregister_callback()
```

thus allowing users to unregister automatically if they invoke ET from the context manager like we do here.

**Review by sanrise — APPROVED** [review]
Overall LGTM, left a couple of comments which should be fairly straightforward to address.

File: torch/profiler/profiler.py (hunk 2 above)
Comment (sanrise) [reply torch/profiler/profiler.py:51]: What do you think about abstracting out the observer to an interface `ITraceObserver` and then provide this `ExecutionTraceObserver` as an implementation of that interface for the community? The functions are very generic anyways (register/unregister/start/stop etc).

File: test/profiler/test_profiler.py [hunk 4]
```diff
@@ -354,34 +377,41 @@ def trace_handler(p):
...
             activities=supported_activities(),
             schedule=torch.profiler.schedule(
                 skip_first=3,
                 wait=1,
                 warmup=1,
-                active=2),
+                active=2,
+                repeat=1),
             on_trace_ready=trace_handler,
+            execution_trace_observer=et,
```
Comment (sanrise) [inline test/profiler/test_profiler.py:396]: Based on my comment below, if we were to extract the observer out to an interface like `ITraceObserver` then this param can just take a list (or single) object satisfying that interface. 

This allows users to just expose any observer (or a list of observers) in the profiler context manager. ExecutionTraceObserver (and hence the Chakra schema) just seem to be a schema our group is interested and settled on, but imo it can vary with the use cases.

**Review by sanrise — COMMENTED** [review]
Added some suggestion to make simple changes to the design to make the idea more generic.

**Comment by aaronenyeshi** [comment]
> We are currently jumping through hoops to collect kineto and ET together and running into numerous reliability issues.

@srinivas212 , what are the reliability issues you are running into? It's important that we keep a high reliability bar for profiling. Are there ways we can collect reliability metrics for Kineto + ET? Any crashing with the two will show up in Kineto monitoring too.

**Comment by briancoutinho** [comment]
@@pytorchbot rebase -s

File: torch/profiler/profiler.py (hunk 2 above)
Comment (briancoutinho) [reply torch/profiler/profiler.py]: done, please see update

File: test/profiler/test_profiler.py (hunk 3 above)
Comment (briancoutinho) [reply test/profiler/test_profiler.py]: done!

File: torch/profiler/profiler.py [hunk 5]
```diff
@@ -65,9 +85,22 @@ class _KinetoProfile:
             then aten::add's module hierarchy is A.B
             Note that this support exist, at the moment, only for TorchScript models
             and not eager mode models.
-
         experimental_config (_ExperimentalConfig) : A set of experimental options
             used by profiler libraries like Kineto. Note, backward compatibility is not guaranteed.
+        execution_trace_observer (ExecutionTraceObserver) : A PyTorch Execution Trace Oobserver object.
```
Comment (aaronenyeshi) [inline torch/profiler/profiler.py:90]: small typo
```suggestion
        execution_trace_observer (ExecutionTraceObserver) : A PyTorch Execution Trace Observer object.
```

File: torch/profiler/profiler.py [hunk 6]
```diff
@@ -65,9 +85,22 @@ class _KinetoProfile:
...
+            When this argument is included the observer start() and stop() will be called for the
+            same time window as PyTorch profiler. The following sample shows how to setup up
+            an Execution Trace observer.
+            ```
+            ...
+            execution_trace_observer=(
+                ExecutionTraceObserver().register_callback("./execution_trac.json")
+            ),
+            ```
+            You can also refer to test_execution_trace_with_kineto() in tests/profiler/test_profier.py.
```
Comment (aaronenyeshi) [inline torch/profiler/profiler.py:102]: small typo
```suggestion
            You can also refer to test_execution_trace_with_kineto() in test/profiler/test_profiler.py.
```

File: torch/profiler/profiler.py [hunk 7]
```diff
@@ -65,9 +85,22 @@ class _KinetoProfile:
...
+            *Note* One can also pass any object sastisfying the _ITraceObserver interface.
```
Comment (aaronenyeshi) [inline torch/profiler/profiler.py:103]:
```suggestion
            *Note* One can also pass any object satisfying the _ITraceObserver interface.
```

File: torch/profiler/profiler.py [hunk 8]
```diff
@@ -418,7 +457,20 @@ class profile(_KinetoProfile):
             and not eager mode models.
         experimental_config (_ExperimentalConfig) : A set of experimental options
             used for Kineto library features. Note, backward compatibility is not guaranteed.
-
+        execution_trace_observer (ExecutionTraceObserver) : A PyTorch Execution Trace Oobserver object.
```
Comment (aaronenyeshi) [inline torch/profiler/profiler.py:460]:
```suggestion
        execution_trace_observer (ExecutionTraceObserver) : A PyTorch Execution Trace Observer object.
```

File: torch/profiler/profiler.py [hunk 9]
```diff
@@ -418,7 +457,20 @@ class profile(_KinetoProfile):
...
+            When this argument is included the observer start() and stop() will be called for the
+            same time window as PyTorch profiler. The following sample shows how to setup up
+            an Execution Trace observer.
+            ```
+            ...
+            execution_trace_observer=(
+                ExecutionTraceObserver().register_callback("./execution_trac.json")
+            ),
+            ```
+            You can also refer to test_execution_trace_with_kineto() in tests/profiler/test_profier.py.
```
Comment (aaronenyeshi) [inline torch/profiler/profiler.py:472]:
```suggestion
            You can also refer to test_execution_trace_with_kineto() in test/profiler/test_profiler.py.
```

File: torch/profiler/profiler.py [hunk 10]
```diff
@@ -418,7 +457,20 @@ class profile(_KinetoProfile):
...
+            *Note* One can also pass any object sastisfying the _ITraceObserver interface.
```
Comment (aaronenyeshi) [inline torch/profiler/profiler.py:473]:
```suggestion
            *Note* One can also pass any object satisfying the _ITraceObserver interface.
```

**Review by aaronenyeshi — APPROVED** [review]
Overall LGTM, we should think about how we will separate reliability metrics between profiler and et.
